common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java

```diff
@@ -0,0 +1,245 @@
+            return resource;
+        }
+
+        StringBuffer prefixBuffer = new StringBuffer();
+        if (isRetryTopic(resource)) {
+            prefixBuffer.append(MixAll.RETRY_GROUP_TOPIC_PREFIX);
+        } else if (isDLQTopic(resource)) {
+            prefixBuffer.append(MixAll.DLQ_GROUP_TOPIC_PREFIX);
+        }
+        prefixBuffer.append(namespace).append(NAMESPACE_SEPARATOR);
+
+        if (resource.startsWith(prefixBuffer.toString())) {
+            return withoutNamespace(resource);
+        }
+
+        return resource;
+    }
+
+    public static String wrapNamespace(String namespace, String resource) {
+        if (StringUtils.isEmpty(namespace) || StringUtils.isEmpty(resource)) {
+            return resource;
+        }
+
+        if (isSystemResource(resource)) {
+            return resource;
+        }
+
+        if (isAlreadyWithNamespace(resource, namespace)) {
+            return resource;
+        }
+
+        StringBuffer strBuffer = new StringBuffer().append(namespace).append(NAMESPACE_SEPARATOR);
+
+        if (isRetryTopic(resource)) {
+            strBuffer.append(resource.substring(RETRY_PREFIX_LENGTH));
+            return strBuffer.insert(0, MixAll.RETRY_GROUP_TOPIC_PREFIX).toString();
+        }
+
+        if (isDLQTopic(resource)) {
+            strBuffer.append(resource.substring(DLQ_PREFIX_LENGTH));
+            return strBuffer.insert(0, MixAll.DLQ_GROUP_TOPIC_PREFIX).toString();
+        }
+
+        return strBuffer.append(resource).toString();
+
+    }
+
+    public static boolean isAlreadyWithNamespace(String resource, String namespace) {
+        if (StringUtils.isEmpty(namespace) || StringUtils.isEmpty(resource) || isSystemResource(resource)) {
+            return false;
+        }
+
+        if (isRetryTopic(resource)) {
+            resource = resource.substring(RETRY_PREFIX_LENGTH);
+        }
+
+        if (isDLQTopic(resource)) {
+            resource = resource.substring(DLQ_PREFIX_LENGTH);
+        }
+
+        return resource.startsWith(namespace + NAMESPACE_SEPARATOR);
+    }
+
+    public static String withNamespaceAndRetry(RemotingCommand request, String consumerGroup) {
+        return wrapNamespaceAndRetry(getNamespaceFromRequest(request), consumerGroup);
+    }
+
+    public static String wrapNamespaceAndRetry(String namespace, String consumerGroup) {
+        if (StringUtils.isEmpty(consumerGroup)) {
+            return null;
+        }
+
+        return new StringBuffer()
+            .append(MixAll.RETRY_GROUP_TOPIC_PREFIX)
+            .append(wrapNamespace(namespace, consumerGroup))
+            .toString();
+    }
+
+    public static String getNamespaceFromRequest(RemotingCommand request) {
+        if (null == request || null == request.getExtFields()) {
+            return null;
+        }
+
+        String namespace;
+
+        switch (request.getCode()) {
+            case RequestCode.SEND_MESSAGE_V2:
+                namespace = request.getExtFields().get("n");
```

Comment:
Here,will you add a field called "n" in the SendMessageRequestHeaderV2 Class File? @lollipopjin

Reply:
Hi，I remove the  redundant parts in NamespaceUtil.java and do some refactors in the new commit.

Reply:
Okay,it's much more better!I will review it later.Thank you!